app/src/main/java/org/schabi/newpipe/local/feed/service/FeedLoadManager.kt

```diff
@@ -0,0 +1,274 @@
+                        error = e
+                    }
+
+                    val request = "${subscriptionEntity.serviceId}:${subscriptionEntity.url}"
+                    val wrapper =
+                        FeedLoadService.RequestException(subscriptionEntity.uid, request, error!!)
+                    return@map Notification.createOnError<FeedUpdateInfo>(wrapper)
+                }
+            }
+            .sequential()
+
+            .observeOn(AndroidSchedulers.mainThread())
+            .doOnNext(NotificationConsumer())
+
+            .observeOn(Schedulers.io())
+            .buffer(BUFFER_COUNT_BEFORE_INSERT)
+            .doOnNext(DatabaseConsumer())
+
+            .subscribeOn(Schedulers.io())
+            .toList()
+            .flatMap { x -> postProcessFeed().toSingleDefault(x.flatten()) }
+    }
+
+    fun cancel() {
+        cancelSignal.set(true)
+    }
+
+    private fun broadcastProgress() {
+        FeedEventManager.postEvent(FeedEventManager.Event.ProgressEvent(currentProgress.get(), maxProgress.get()))
+    }
+
+    /**
+     * Keep the feed and the stream tables small
+     * to reduce loading times when trying to display the feed.
+     * <br>
+     * Remove streams from the feed which are older than [FeedDatabaseManager.FEED_OLDEST_ALLOWED_DATE].
+     * Remove streams from the database which are not linked / used by any table.
+     */
+    private fun postProcessFeed() = Completable.fromRunnable {
+        FeedEventManager.postEvent(FeedEventManager.Event.ProgressEvent(R.string.feed_processing_message))
+        feedDatabaseManager.removeOrphansOrOlderStreams()
+
+        FeedEventManager.postEvent(FeedEventManager.Event.SuccessResultEvent(feedResultsHolder.itemsErrors))
+    }.doOnSubscribe {
+        currentProgress.set(-1)
+        maxProgress.set(-1)
+
+        notificationUpdater.onNext(context.getString(R.string.feed_processing_message))
+        FeedEventManager.postEvent(FeedEventManager.Event.ProgressEvent(R.string.feed_processing_message))
+    }.subscribeOn(Schedulers.io())
+
+    private inner class NotificationConsumer : Consumer<Notification<FeedUpdateInfo>> {
+        override fun accept(item: Notification<FeedUpdateInfo>) {
+            currentProgress.incrementAndGet()
+            notificationUpdater.onNext(item.value?.name.orEmpty())
+
+            broadcastProgress()
+        }
+    }
+
+    private inner class DatabaseConsumer : Consumer<List<Notification<FeedUpdateInfo>>> {
+
+        override fun accept(list: List<Notification<FeedUpdateInfo>>) {
+            feedDatabaseManager.database().runInTransaction {
+                for (notification in list) {
+                    when {
+                        notification.isOnNext -> {
+                            val subscriptionId = notification.value.uid
+                            val info = notification.value.listInfo
+
+                            notification.value.newStreamsCount = countNewStreams(info.relatedItems)
```

Comment:
*(to solve the above problem(s), 2)*
I would remove the `newStreamsCount` altogether and instead calculate the `newStreams` array here